### ISSUE TITLE
"bool" cannot be assigned to type "Literal[True]"

### ISSUE THREAD

**Issue by dbrattli** [microsoft/pyright#2867]
**Describe the bug**

The included code example gives error: `"bool" cannot be assigned to type "Literal[True]"`. This might be a bug with pyright. Ref: https://github.com/microsoft/pyright/issues/2796

**Screenshots or Code**

```python
from typing import Callable, TypeVar, Generic, Tuple

_T = TypeVar("_T")

class IAsyncContext(Generic[_T]):
    pass

Async = Callable[[IAsyncContext[_T]], None]

def Return(value: _T) -> Async[_T]:
    def ret(ctx: IAsyncContext[_T]) -> None:
        pass

    return ret

def arrow_3() -> Async[Tuple[None, bool]]:
    return Return((None, True))
```

**VS Code extension or command-line**

Vscode insiders: pyright: v1.1.209

**Comment by erictraut** [comment]
Here's a simplified version that exhibits the problem:

```python
from typing import Callable, TypeVar, Tuple

_T = TypeVar("_T")

def func1(value: _T) -> Callable[[_T], None]:
    ...

x: Callable[[Tuple[bool]], None] = func1((True,))
```

This is proving to be a tough one to fix. It's triggering several heuristics in the TypeVar solver and bidirectional inference logic relating to callable parameters (which are implicitly contravariant) and types with literals. If I modify the heuristics to accommodate this case, other cases break. While mypy handles the above case without error, it fails on some of the other cases that these heuristics are designed to accommodate.

I'll need to think about it more and see if I can come up with even more complex heuristics that accommodate the full set of test cases that exercise these heuristics. So far, my attempts to do so have failed.

**Comment by erictraut** [comment]
OK, I've found a way to accommodate all existing cases as well as this one. This will be addressed in the next release.

**Comment by dbrattli** [comment]
Thank you for fixing. PS: I'm currently working on Python support for [Fable](https://fable.io/) i.e transpiling F# to Python, so that's why I'm starting to discover a bit unusual cases. The type systems are not fully compatible which is a challenge, but for most it works really well.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.210, which I just published. It will also be included in the next release of pylance.
